datajoint_pacman/pacman_processing.py

```diff
@@ -1,18 +1,69 @@
 import datajoint as dj
-from churchland_pipeline_python import lab, acquisition, processing
-from churchland_pipeline_python.utilities import datasync
+import numpy as np
+import itertools, inspect
+from churchland_pipeline_python import lab, acquisition, processing, equipment
+from churchland_pipeline_python.utilities import datasync, datajointutils as dju
 from . import pacman_acquisition
 from brpylib import NsxFile, brpylib_ver
 from datetime import datetime
 
 schema = dj.schema('churchland_analyses_pacman_processing')
 
 @schema
-class AlignmentState(dj.Lookup):
+class AlignmentParams(dj.Manual):
     definition = """
     # Task state IDs used to align trials
+    -> pacman_acquisition.Behavior
+    alignment_params_id: tinyint unsigned
+    ---
     -> pacman_acquisition.TaskState
+    alignment_max_lag = 0.2: decimal(4,3) # maximum allowable lag (s)
+    """
+    
+    @classmethod
+    def populate(self, 
+        behavior_rel: pacman_acquisition.Behavior=pacman_acquisition.Behavior(), 
+        task_state_rel: pacman_acquisition.TaskState=(pacman_acquisition.TaskState & {'task_state_name': 'InTarget'}), 
+        max_lag: int=0.2):
+
+        # check inputs
+        assert isinstance(behavior_rel, pacman_acquisition.Behavior), 'Unrecognized behavior table'
+        assert isinstance(task_state_rel, pacman_acquisition.TaskState), 'Unrecognized task state table'
+
+        # construct "key source" from join of behavior and task state tables
+        key_source = (behavior_rel * task_state_rel) - (self & {'alignment_max_lag': max_lag})
+
+        behavior_source = behavior_rel & key_source.proj()
+        task_state_source = task_state_rel & key_source.proj()
+
+        # insert task state for every behavior
+        for beh_key, task_state_key in itertools.product(behavior_source.fetch('KEY'), task_state_source.fetch('KEY')):
+
+            # get filter params ID
+            if not self & beh_key:
+                new_param_id = 0
+            else:
+                all_param_id = (self & beh_key).fetch('alignment_params_id')
+                new_param_id = next(i for i in range(2+max(all_param_id)) if i not in all_param_id)
+
+            self.insert1(dict(**beh_key, alignment_params_id=new_param_id, **task_state_key, alignment_max_lag=max_lag))
+
+
+@schema
+class BehaviorBlock(dj.Manual):
+    definition = """
+    # Set of save tags and behavioral recording parameters for conducting analyses
+    -> pacman_acquisition.Behavior
+    behavior_block_id: tinyint unsigned # block ID
+    ---
+    -> pacman_acquisition.ArmPosture
     """
+    
+    class SaveTag(dj.Part):
+        definition = """
+        -> master
+        -> pacman_acquisition.Behavior.SaveTag
+        """
 
 @schema
 class EphysTrialStart(dj.Imported):
```

Comment:
`key_source` is for each Trial, but it seems that the trials are inserted together for each session. Shall we modify the key_source to the level of session?